### PR TITLE
Refactor hammer as a library

### PR DESCRIPTION
**Pull request by gdbelvin**
This PR allows hammer operations to be used in different contexts.
It also slightly simplifies the channel logic that was used in the hammer.

### CODE REVIEW

File: core/client/hammer/hammer.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+		appID:   fmt.Sprintf("hammer_%v", time.Now().Format("2006-01-02/15:04:05")),
+		config:  config,
+
+		signers:        []*tink.KeysetHandle{keyset},
+		authorizedKeys: authorizedKeys.Keyset(),
+	}, nil
+}
+
+// Run runs operationCount with up to maxWorkers over ramp duration.
+func (h *Hammer) Run(ctx context.Context, operationCount, maxWorkers int, ramp time.Duration) {
```
Comment (pavelkalinnikov) [inline core/client/hammer/hammer.go]: Maybe make this whole thing (down to `worker`) `ctx`-cancellable?

File: core/client/hammer/hammer.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	go func() {
+		for i := 0; i < count; i++ {
+			select {
+			case <-ctx.Done():
+				close(jobs)
+				return
+			case jobs <- i:
+			}
+		}
+		close(jobs)
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply core/client/hammer/hammer.go]: Good idea

File: core/client/hammer/hammer.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+			rate := totalRequests / time.Since(startTime).Seconds()
+			glog.Infof("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
+		}
+	}()
+
+	// Collect outputs until output closes.
+	for range outputs {
+		totalRequests++
+	}
+	refresh.Stop()
```
[the remark this answers is not in the document]
Comment (gdbelvin) [reply core/client/hammer/hammer.go]: Thanks for catching that

File: core/client/hammer/hammer.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+	// Print stats asyncronously until refresh closes.
+	refresh := time.NewTicker(period)
+	stop := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-refresh.C:
+				rate := totalRequests / time.Since(startTime).Seconds()
+				glog.Infof("QPS: %v, \tRequests: %v, \tWorkers: %v", rate, totalRequests, h.workers)
+			case <-stop:
```
Comment (pavelkalinnikov) [inline core/client/hammer/hammer.go]: Shouldn't you write to `stop` channel to make this `case` happen?

File: core/client/hammer/hammer.go (hunk 1 above)
Comment (pavelkalinnikov) [reply core/client/hammer/hammer.go]: You use `case <-ctx.Done():` clause below in `gen` function to make it early cancellable, right? I was suggesting to maybe do the same in `launchWorkers` and `printStats`. But after second thought I see that those will actually terminate after `gen` gets cancelled, as they both depend on it through channels.
So you can ignore this comment.

File: core/client/hammer/hammer.go (hunk 4 above)
Comment (gdbelvin) [reply core/client/hammer/hammer.go]: `close(stop)` causes the channel return immediately

**Pull request merged**